### PR TITLE
MAYA-123720 - As a user, I'd like Display Layer membership to survive the 'Edit As Maya Data' workflow

### PR DESCRIPTION
**Pull request by vlasovi**

### CODE REVIEW

File: lib/mayaUsd/ufe/Utils.h
Comment (seando-adsk) [inline lib/mayaUsd/ufe/Utils.h:241]: Why not store the MFnDisplayLayer so you don't need the extra check/object create in the processItem()?

File: lib/mayaUsd/ufe/Utils.h
Comment (seando-adsk) [inline lib/mayaUsd/ufe/Utils.h:234]: Do we need two different classes? How about a single class that can handle both cases (Maya->USD and USD->Maya). Just have two sections of methods for each case. And I guess you were thinking that in the future we could add other types of extras (now we have DisplayLayers)? The names "Pull/Push" are specific to your operation, but we'll add this for duplicate usd prim. Can we have a generic name?

File: lib/mayaUsd/ufe/Utils.h
Comment (seando-adsk) [inline lib/mayaUsd/ufe/Utils.h:238]: These methods need comments as to what they are doing.

File: lib/mayaUsd/ufe/Utils.h
Comment (vlasovi) [reply lib/mayaUsd/ufe/Utils.h:241]: @seando-adsk I cannot put MFnDisplayLayer in a container because its operator= as well as its copy constructor are declared private and not accessible. Putting a pointer to MFnDisplayLayer is basically the same as using MObject

File: lib/mayaUsd/ufe/Utils.h
Comment (vlasovi) [reply lib/mayaUsd/ufe/Utils.h:234]: @seando-adsk The two classes share neither data members nor methods, so I don't see a reason to merge them. As for renaming, would ReplicateExtrasToUSD/ReplicateExtrasFromUSD wouldwork for you?

Comment (seando-adsk) [reply lib/mayaUsd/ufe/Utils.h:234]: Yes those names are better.

**Review by seando-adsk — CHANGES_REQUESTED** [review]
As a general comment you need some method documentation and unit tests.

**Review by seando-adsk — DISMISSED** [review]
Looks good thanks.

**Comment by seando-adsk** [comment]
I'll wait until @ppt-adsk has had a chance to review this before merging.

File: lib/mayaUsd/fileio/primUpdaterContext.h
Comment (ppt-adsk) [inline lib/mayaUsd/fileio/primUpdaterContext.h:71]: These mutable data members feel uncomfortable.  Does the following make any sense:
- initRecursive() writes into the map.  Can we transform it into a static creator method?
- processItem() and finalize() read from the map.  They can be made const methods.
That way, you can pass in the ReplicateExtras objects as const into the context, make them const data members, and provide const accessors to them, keeping the PrimUpdaterContext read-only.

Hope I didn't skip over some crucial detail...

**Review by ppt-adsk — COMMENTED** [review]
Great work, see my suggestion and let me know if it makes sense.

File: lib/mayaUsd/ufe/Utils.cpp
Comment (vlasovi) [reply lib/mayaUsd/ufe/Utils.cpp:915]: Yeah, that was my thinking as well.

File: lib/mayaUsd/fileio/primUpdaterContext.h
Comment (vlasovi) [reply lib/mayaUsd/fileio/primUpdaterContext.h:71]: Your plan will not fully work because ReplicateExtrasToUSD::processItem writes into the map.

Comment (ppt-adsk) [reply lib/mayaUsd/fileio/primUpdaterContext.h:71]: O.K., FromUSD works, ToUSD does not :)  Let me think about it some more.

Comment (ppt-adsk) [reply lib/mayaUsd/fileio/primUpdaterContext.h:71]: Here's a tweak to my suggestion to get it to work: everything as per above, but ReplicateExtrasToUSD::processItem() does a const cast if it needs to write into the map.  Alternately, the map is mutable.  I think this is acceptable, as the fact that processItem() is writing into the map is an invisible implementation detail.  In that way, the prim updater context can appear to be conceptually read-only.  What do you think?

**Review by ppt-adsk — APPROVED** [review]
Thanks for the changes, looks good!

**Pull request merged**
